### ISSUE TITLE
Build fails on Debian Wheezy 64bit with backported Kernel 3.14

### ISSUE THREAD

**Issue by ChojinDSL** [markfasheh/duperemove#44]
I'm trying to build duperemove on a debian wheezy Server using kernel 3.14. 

Whenever I try to build, I get the following error:

###########################################
make
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.10-dev\"  -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include   -rdynamic -c duperemove.c -o duperemove.o  -lglib-2.0  
duperemove.c: In function ‘main’:
duperemove.c:290:2: warning: implicit declaration of function ‘g_get_num_processors’ [-Wimplicit-function-declaration]
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.10-dev\"  -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include   -rdynamic -c hash-tree.c -o hash-tree.o  -lglib-2.0  
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.10-dev\"  -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include   -rdynamic -c results-tree.c -o results-tree.o  -lglib-2.0  
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.10-dev\"  -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include   -rdynamic -c rbtree.c -o rbtree.o  -lglib-2.0  
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.10-dev\"  -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include   -rdynamic -c dedupe.c -o dedupe.o  -lglib-2.0  
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.10-dev\"  -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include   -rdynamic -c filerec.c -o filerec.o  -lglib-2.0  
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.10-dev\"  -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include   -rdynamic -c btrfs-util.c -o btrfs-util.o  -lglib-2.0  
btrfs-util.c:27:25: fatal error: linux/btrfs.h: Datei oder Verzeichnis nicht gefunden
compilation terminated.
make: *** [btrfs-util.o] Fehler 1
##################################################

That part with "fatal error: linux/btrfs.h:" says "File or Directory not found"

I'm guessing it's probably something minor, like a missing library, or it's looking in the wrong directory. Perhaps a symlink is missing. 
The odd thing is, btrfs.h exists. Its located in /usr/src/linux-headers-3.14-0.bpo.1-common/include/linux/btrfs.h

Does anyone have an idea on what I'm missing?

**Comment by ChojinDSL** [comment]
Strange, just tested in on my Debian Jessie Laptop with Kernel 3.12, and there it builds without issue.
(I'm currently booted into 3.12 on my laptop, but 3.14 3.16 and 3.17 are also installed)

Are there some libs that are too outdated on Debian Wheezy perhaps?

On my Debian Wheezy server I have btrfs-tools version 3.14.1
On my Debian Jessie laptop I have btrfs-tools version 3.17

**Comment by markfasheh** [comment]
Where are you getting linux/btrfs.h from on the Debian Jessie laptop? Basically it you're missing the btrfs kernel header file on the machine that can't compile. You might need a kernel-headers or kernel-devel package (I'm not familiar with Debian). Also, I am concerned about:

duperemove.c: In function ‘main’:
duperemove.c:290:2: warning: implicit declaration of function ‘g_get_num_processors’ [-Wimplicit-function-declaration]

Could that indicate you have an old or not existing version of glib2?

**Comment by ChojinDSL** [comment]
On Debian Jessie, linux/btrfs.h is located in /usr/src/linux-headers-<$kernel-version>
I'm guessing its using the one of the current kernel I'm running.

But I think it's probably due to an old version of glib2. On my wheezy (stable) Server glib2 is version 2.32.4 and on the Jessie Laptop its 2.42.

Debian is known for its conservative package versions. At least in the stable distro, in this case wheezy. 

But that would make sense, since the problem compiling it, doesn't show up on more up to date distros. 
But good to know. Guess I'll just have to wait with my servers until Jessie hits stable, which shouldn't be too far off. 

Curious though, why the dependency on glib2? Isn't that a GTK library?

Anyway, thanks for your help.

**Comment by RX14** [comment]
I'm getting a similar issue, without the glib problems:
```
root@RX-FILE-V1:~/duperemove# make
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.09\"  -I/usr/include/glib-2.0 -I/usr/lib/i386-linux-gnu/glib-2.0/include   -rdynamic -c duperemove.c -o duperemove.o -L/lib/i386-linux-gnu -lgcrypt -lglib-2.0  
duperemove.c: In function ‘populate_hash_tree’:
duperemove.c:543:3: warning: implicit declaration of function ‘g_get_num_processors’ [-Wimplicit-function-declaration]
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.09\"  -I/usr/include/glib-2.0 -I/usr/lib/i386-linux-gnu/glib-2.0/include   -rdynamic -c hash-tree.c -o hash-tree.o -L/lib/i386-linux-gnu -lgcrypt -lglib-2.0  
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.09\"  -I/usr/include/glib-2.0 -I/usr/lib/i386-linux-gnu/glib-2.0/include   -rdynamic -c results-tree.c -o results-tree.o -L/lib/i386-linux-gnu -lgcrypt -lglib-2.0  
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.09\"  -I/usr/include/glib-2.0 -I/usr/lib/i386-linux-gnu/glib-2.0/include   -rdynamic -c rbtree.c -o rbtree.o -L/lib/i386-linux-gnu -lgcrypt -lglib-2.0  
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.09\"  -I/usr/include/glib-2.0 -I/usr/lib/i386-linux-gnu/glib-2.0/include   -rdynamic -c dedupe.c -o dedupe.o -L/lib/i386-linux-gnu -lgcrypt -lglib-2.0  
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.09\"  -I/usr/include/glib-2.0 -I/usr/lib/i386-linux-gnu/glib-2.0/include   -rdynamic -c filerec.c -o filerec.o -L/lib/i386-linux-gnu -lgcrypt -lglib-2.0  
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.09\"  -I/usr/include/glib-2.0 -I/usr/lib/i386-linux-gnu/glib-2.0/include   -rdynamic -c btrfs-util.c -o btrfs-util.o -L/lib/i386-linux-gnu -lgcrypt -lglib-2.0  
btrfs-util.c:27:25: fatal error: linux/btrfs.h: No such file or directory
compilation terminated.
make: *** [btrfs-util.o] Error 1
```

`btrfs.h` exists in `/usr/src/linux-headers-3.18.9-rx14/include/uapi/linux/` but it seems it is not being found.

I'm not experiences with C and stuff, but i thought i would try `LIBRARY_FLAGS="-I/usr/src/linux-headers-3.18.9-rx14/include/" make`  but it shows this:
```
gcc -Wall -ggdb -D_FILE_OFFSET_BITS=64 -DVERSTRING=\"v0.09\"  -I/usr/include/glib-2.0 -I/usr/lib/i386-linux-gnu/glib-2.0/include   -rdynamic -c btrfs-util.c -o btrfs-util.o -I/usr/src/linux-headers-3.18.9-rx14/include/ -L/lib/i386-linux-gnu -lgcrypt -lglib-2.0  
In file included from /usr/src/linux-headers-3.18.9-rx14/include/uapi/linux/btrfs.h:21:0,
                 from /usr/src/linux-headers-3.18.9-rx14/include/linux/btrfs.h:4,
                 from btrfs-util.c:27:
/usr/src/linux-headers-3.18.9-rx14/include/linux/types.h:14:26: error: conflicting types for ‘fd_set’
In file included from /usr/include/i386-linux-gnu/sys/types.h:220:0,
                 from /usr/include/stdlib.h:320,
                 from btrfs-util.c:19:
/usr/include/i386-linux-gnu/sys/select.h:78:5: note: previous declaration of ‘fd_set’ was here
In file included from /usr/src/linux-headers-3.18.9-rx14/include/uapi/linux/btrfs.h:21:0,
                 from /usr/src/linux-headers-3.18.9-rx14/include/linux/btrfs.h:4,
                 from btrfs-util.c:27:
/usr/src/linux-headers-3.18.9-rx14/include/linux/types.h:15:25: error: conflicting types for ‘dev_t’
In file included from /usr/include/stdlib.h:320:0,
                 from btrfs-util.c:19:
/usr/include/i386-linux-gnu/sys/types.h:61:17: note: previous declaration of ‘dev_t’ was here
In file included from /usr/src/linux-headers-3.18.9-rx14/include/uapi/linux/btrfs.h:21:0,
                 from /usr/src/linux-headers-3.18.9-rx14/include/linux/btrfs.h:4,
                 from btrfs-util.c:27:
/usr/src/linux-headers-3.18.9-rx14/include/linux/types.h:16:25: error: conflicting types for ‘ino_t’
In file included from /usr/include/stdlib.h:320:0,
                 from btrfs-util.c:19:
/usr/include/i386-linux-gnu/sys/types.h:51:19: note: previous declaration of ‘ino_t’ was here
In file included from /usr/src/linux-headers-3.18.9-rx14/include/uapi/linux/btrfs.h:21:0,
                 from /usr/src/linux-headers-3.18.9-rx14/include/linux/btrfs.h:4,
                 from btrfs-util.c:27:
/usr/src/linux-headers-3.18.9-rx14/include/linux/types.h:17:26: error: conflicting types for ‘mode_t’
In file included from /usr/include/stdlib.h:320:0,
                 from btrfs-util.c:19:
/usr/include/i386-linux-gnu/sys/types.h:71:18: note: previous declaration of ‘mode_t’ was here
In file included from /usr/src/linux-headers-3.18.9-rx14/include/uapi/linux/btrfs.h:21:0,
                 from /usr/src/linux-headers-3.18.9-rx14/include/linux/btrfs.h:4,
                 from btrfs-util.c:27:
/usr/src/linux-headers-3.18.9-rx14/include/linux/types.h:20:25: error: conflicting types for ‘off_t’
In file included from btrfs-util.c:18:0:
/usr/include/unistd.h:246:19: note: previous declaration of ‘off_t’ was here
In file included from /usr/src/linux-headers-3.18.9-rx14/include/uapi/linux/btrfs.h:21:0,
                 from /usr/src/linux-headers-3.18.9-rx14/include/linux/btrfs.h:4,
                 from btrfs-util.c:27:
/usr/src/linux-headers-3.18.9-rx14/include/linux/types.h:25:26: error: conflicting types for ‘timer_t’
In file included from /usr/include/i386-linux-gnu/sys/types.h:133:0,
                 from /usr/include/stdlib.h:320,
                 from btrfs-util.c:19:
/usr/include/time.h:104:19: note: previous declaration of ‘timer_t’ was here
In file included from /usr/src/linux-headers-3.18.9-rx14/include/uapi/linux/btrfs.h:21:0,
                 from /usr/src/linux-headers-3.18.9-rx14/include/linux/btrfs.h:4,
                 from btrfs-util.c:27:
/usr/src/linux-headers-3.18.9-rx14/include/linux/types.h:36:24: error: conflicting types for ‘uintptr_t’
In file included from /usr/lib/gcc/i486-linux-gnu/4.7/include/stdint.h:3:0,
                 from btrfs-util.c:20:
/usr/include/stdint.h:129:23: note: previous declaration of ‘uintptr_t’ was here
In file included from /usr/src/linux-headers-3.18.9-rx14/include/uapi/linux/btrfs.h:21:0,
                 from /usr/src/linux-headers-3.18.9-rx14/include/linux/btrfs.h:4,
                 from btrfs-util.c:27:
/usr/src/linux-headers-3.18.9-rx14/include/linux/types.h:134:23: error: conflicting types for ‘blkcnt_t’
In file included from /usr/include/stdlib.h:320:0,
                 from btrfs-util.c:19:
/usr/include/i386-linux-gnu/sys/types.h:249:22: note: previous declaration of ‘blkcnt_t’ was here
In file included from btrfs-util.c:30:0:
debug.h:66:0: warning: "compiletime_assert" redefined [enabled by default]
In file included from /usr/src/linux-headers-3.18.9-rx14/include/uapi/linux/stddef.h:1:0,
                 from /usr/src/linux-headers-3.18.9-rx14/include/linux/stddef.h:4,
                 from /usr/include/linux/posix_types.h:4,
                 from /usr/src/linux-headers-3.18.9-rx14/include/uapi/linux/types.h:13,
                 from /usr/src/linux-headers-3.18.9-rx14/include/linux/types.h:5,
                 from /usr/src/linux-headers-3.18.9-rx14/include/uapi/linux/btrfs.h:21,
                 from /usr/src/linux-headers-3.18.9-rx14/include/linux/btrfs.h:4,
                 from btrfs-util.c:27:
/usr/src/linux-headers-3.18.9-rx14/include/linux/compiler.h:362:0: note: this is the location of the previous definition
make: *** [btrfs-util.o] Error 1
```
